### PR TITLE
Article/frontend good practices

### PR DESCRIPTION
**Pull request by jsconan**
Document a list of good practices for day to day work on frontend.

### CODE REVIEW

File: articles/Frontend/Architecture/good-practices.md [hunk 1]
```diff
@@ -0,0 +1,1724 @@
...
+## Testing
+
+### Prefer design by coding: TDD / BDD
+When writing code, a developer must check the works is going the right way.
+Usually, some manual checks are made aside. But it may become more and more
+time consuming as long as the progress is made. So it is better to automate
+somehow those manual checks. And on convenient way to do so is to write
+unit tests.
+
+A method well describes the process, it is called [Test Driven Development](https://en.wikipedia.org/wiki/Test-driven_development).
```
Comment (krampstudio) [inline articles/Frontend/Architecture/good-practices.md]: Can you please rephrase

File: articles/Frontend/Architecture/good-practices.md [hunk 2]
```diff
@@ -0,0 +1,1724 @@
...
+The principle is to write tests aside at the same period of time you are
```
Comment (krampstudio) [inline articles/Frontend/Architecture/good-practices.md]: not at the same time, it's before (start with the test then make your implementation fulfil it)

File: articles/Frontend/Architecture/good-practices.md [hunk 3]
```diff
@@ -0,0 +1,1724 @@
...
+button('.fixture', {label: 'Ok'})
+    .on('click', () => { ... });
+```
+
+#### Resources
+- [Coding guide: General rules](coding-guide.md#general-rules)
+
+## Testing
+
+### Prefer design by coding: TDD / BDD
```
Comment (krampstudio) [inline articles/Frontend/Architecture/good-practices.md]: I'm not sure to get the link with BDD

File: articles/Frontend/Architecture/good-practices.md [hunk 4]
```diff
@@ -0,0 +1,1724 @@
...
+cleaned between each iteration. However, when the tests are ran in parallel
+there is a big chance they will conflict, and will produce inconsistent results.
+
+Having inconsistent tests might be the symptom of concurrency issue within
+the test suite.
+
+To prevent any conflict, and to make sure each unit test is properly scoped
+this is a good habit to reserve one unique markup for each different test. 
+Keeping the same markup for a serie of tests is still legit however, as they 
+will run one after the other. The requirement being to clean the markup
```
Comment (krampstudio) [inline articles/Frontend/Architecture/good-practices.md]: You can actually reuse the same fixtures between tests. 
After each test the fixture element is removed and created again, so if you all your tests relies on the same structure you can safely reuse the fixture in the next test.

Comment (jsconan) [reply articles/Frontend/Architecture/good-practices.md]: Yes indeed, the fixture is cleaned by the test engine. However, it may also run the tests in parallel, and in that case the tests will conflict if they expect incompatible changes.

File: articles/Frontend/Architecture/good-practices.md [hunk 5]
```diff
@@ -0,0 +1,1724 @@
...
+});
+
+tabs.on('activate', id => console.log('tab activated', id));
+
+// the action will be immediately performed
+tabs.activate('tab2');
+```
+
+#### Improved example: defer the action to event
+The previous example can be improved using action event instead of immediate
```
[the remark this answers is not in the document]
Comment (jsconan) [reply articles/Frontend/Architecture/good-practices.md]: indeed, good catch. I augmented the explanation based on your suggestion.

File: articles/Frontend/Architecture/good-practices.md (hunk 4 above)
Comment (krampstudio) [reply articles/Frontend/Architecture/good-practices.md]: as far as I know QUnit guarantee the DOM of the fixture will be the defined DOM for each and every test case (and doesn't run the tests in parallel, can be in a different order but it doesn't support parallel tests)

File: articles/Frontend/Architecture/good-practices.md [hunk 6]
```diff
@@ -1334,16 +1360,16 @@ button('.fixture', {label: 'Ok'})
...
 time consuming as long as the progress is made. So it is better to automate
-somehow those manual checks. And on convenient way to do so is to write
-unit tests.
+somehow those manual checks. And a convenient way to do so is to write unit 
+tests.
 
-A method well describes the process, it is called [Test Driven Development](https://en.wikipedia.org/wiki/Test-driven_development).
-The principle is to write tests aside at the same period of time you are
-implementing. The best being to start writing test, then implement the feature.
+In order to ease the process, and in order to improve the quality of the
```
Comment (krampstudio) [inline articles/Frontend/Architecture/good-practices.md]: I'd rephrase it also to ` In order to ease the process and improve the`

File: articles/Frontend/Architecture/good-practices.md [hunk 7]
```diff
@@ -1128,7 +1144,12 @@ tabs.activate('tab2');
 
 #### Improved example: defer the action to event
 The previous example can be improved using action event instead of immediate 
-action. 
+action. The benefit being you can now prevent the tab to be activated.
+
+However, please keep in mind this is not an universal solution. Depending on
+what you intend it might not be the better approach. Once again, ask yourself
```
Comment (krampstudio) [inline articles/Frontend/Architecture/good-practices.md]: `better` -> `best`

File: articles/Frontend/Architecture/good-practices.md (hunk 4 above)
Comment (jsconan) [reply articles/Frontend/Architecture/good-practices.md]: Ok, my bad.
I rephrased the section

File: articles/Frontend/Architecture/good-practices.md (hunk 7 above)
Comment (jsconan) [reply articles/Frontend/Architecture/good-practices.md]: good catch

File: articles/Frontend/Good Practices/modules.md [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+In that case it is far better to start and stop the process when needed.
+
+That being said, we must generally consider the following:
+- Avoid to directly execute code from a module, unless this is needed to build
+some structure and the code is fast enough to not have negative impact on the
+page bootstrap.
+- Avoid to register events from the module, prefer to implement an API to start
+and stop the event listening when needed.
+- For workers or bootstrap modules, having auto executed code is legit.
+- Registering adapters is also ok since the purpose is to load not to start a
```
Comment (krampstudio) [inline articles/Frontend/Good Practices/modules.md]: I'm not sure to understand that sentence

File: articles/Frontend/Good Practices/styling.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		- [Bad example: building class names with parent selector](#bad-example-building-class-names-with-parent-selector)
+		- [Good example: finite class names and inheritance with parent selector](#good-example-finite-class-names-and-inheritance-with-parent-selector)
+	- [Use dedicated class names](#use-dedicated-class-names)
+		- [Bad example: unspecific class names](#bad-example-unspecific-class-names)
+		- [Good example: specific class names](#good-example-specific-class-names)
+
+<!-- /TOC -->
+
+## Use explicit class names
+SASS and LESS are providing some useful shortcuts, but could also lead to
```
Comment (krampstudio) [inline articles/Frontend/Good Practices/styling.md]: no need to mention `less`

File: articles/Frontend/Good Practices/styling.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+## Use dedicated class names
+Always prefer specific and well defined class name to match a component part.
+Using generic name could lead to conflicts as they could be broadly used for
+different purposes. Unless, of course, the intent is to have a generic behavior,
+like `disabled` or `hidden`. But when we talk about scope class, it is better
+to rely on unique and self-explaining names. Please avoid also verbose styling,
+like Bootstrap use to do. This has the same downside as hardcoding the style
+within the markup. It couples hard meaning to the design, and this is not easy
+to apply proper design later on.
```
Comment (krampstudio) [inline articles/Frontend/Good Practices/styling.md]: No need to mention Bootstrap. Just say the utility classes that contains the value in the name create coupling.

**Comment by jsconan** [comment]
@krampstudio I finally had the time to apply the changes to this long-running PR.
Please review when you have time ;)

**Review by krampstudio — APPROVED** [review]
I didn't check the details this time, but it looks good overall
